misc/python/materialize/mzcompose/composition.py

```diff
@@ -263,48 +268,49 @@ def invoke(
             ("--project-name", self.project_name) if self.project_name else ()
         )
 
-        # Emit the munged configuration to a temporary file so that we can later
-        # pass it to Docker Compose.
-        file = TemporaryFile(mode="w")
-        os.set_inheritable(file.fileno(), True)
-        yaml.dump(self.compose, file)
-        file.flush()
-
-        ret = None
-        for retry in range(1, max_tries + 1):
-            file.seek(0)
-            try:
-                ret = subprocess.run(
-                    [
-                        "docker",
-                        "compose",
-                        f"-f/dev/fd/{file.fileno()}",
-                        "--project-directory",
-                        self.path,
-                        *project_name_args,
-                        *args,
-                    ],
-                    close_fds=False,
-                    check=check,
-                    stdout=stdout,
-                    stderr=stderr,
-                    input=stdin,
-                    text=True,
-                    bufsize=1,
-                )
-            except subprocess.CalledProcessError as e:
-                if e.stdout:
-                    print(e.stdout)
-
-                if retry < max_tries:
-                    print("Retrying ...")
-                    time.sleep(1)
-                    continue
-                else:
-                    raise UIError(
-                        f"running docker compose failed (exit status {e.returncode})"
+        with self.lock:
+            # Emit the munged configuration to a temporary file so that we can later
+            # pass it to Docker Compose.
+            self.file.seek(0)
+            self.file.truncate()
+            yaml.dump(self.compose, self.file)
+            self.file.flush()
+
+            ret = None
+            for retry in range(1, max_tries + 1):
+                self.file.seek(0)
+                try:
+                    ret = subprocess.run(
+                        [
+                            "docker",
+                            "compose",
+                            f"-f/dev/fd/{self.file.fileno()}",
+                            "--project-directory",
+                            self.path,
+                            *project_name_args,
+                            *args,
+                        ],
+                        close_fds=False,
+                        check=check,
+                        stdout=stdout,
+                        stderr=stderr,
+                        input=stdin,
+                        text=True,
+                        bufsize=1,
                     )
-            break
+                except subprocess.CalledProcessError as e:
+                    if e.stdout:
+                        print(e.stdout)
```

Comment:
What about `stderr`?

Reply:
I didn't change that part, so not sure what the original motivation was.

Reply:
I am not 100% certain, but I think docker compose writes everything to stdout, including errors and such.